// generator-jshint-config/generators/app/index.js
var Generator = require('yeoman-generator');
var prompts = require('./util/prompts.js');
var getSettings = require('./util/getsettings.js');
var chalk = require('chalk');
var yosay = require('yosay');

module.exports = class extends Generator {
  initializing() {
    this.dependencies = ['jshint'];
  }

  prompting() {
    this.log(yosay(
      `Welcome to the ${chalk.green('jshint-config')}!`
    ));

    return this.prompt(prompts).then(answers => {
      this.answers = getSettings(answers);
      this.jsPath = answers.jsFiles;
    });
  }

  writing() {
    this.fs.copyTpl(
      this.templatePath('.jshintrc'),
      this.destinationPath('.jshintrc'),
      this.answers
    );

    this.fs.extendJSON(
      this.destinationPath('package.json'),
      {
        scripts: {
          jshint: `jshint --config .jshintrc "${this.answers.jsFiles}"`
        }
      }
    );

    this.log(chalk.green(`Adding useful scripts to a ${chalk.white('package.json')} file.`));
    this.log(chalk.green(`Run ${chalk.bold('jshint')} to lint your JS files.`));
  }

  install() {
    this.yarnInstall(this.dependencies, {dev: true});
    this.installDependencies({
      yarn: true,
      npm: true,
      bower: false
    });
  }

  end() {
    this.log(chalk.green('... We are done!'));
    this.log(chalk.green(`Thanks for using ${chalk.bold('jshint-config')}. Hava a nice day!`));
  }
}
